components/webln.js

```diff
@@ -0,0 +1,106 @@
+import { createContext, createRef, useContext, useEffect, useImperativeHandle, useState } from 'react'
+
+const WebLNContext = createContext({})
+export const WebLNContextRef = createRef()
+
+const fetchWebLNProvider = async () => {
+  // sync provider from local storage
+  // TODO check if there is a better, more secure option
+  const configJson = window.localStorage.getItem('bc:config')
+  if (configJson) {
+    const WebLNProviders = await import('@getalby/bitcoin-connect').then((mod) => mod.WebLNProviders)
+    const config = JSON.parse(configJson)
+    if (config.connectorType === 'lnbits') return new WebLNProviders.LnbitsWebLNProvider(config.lnbitsInstanceUrl, config.lnbitsAdminKey)
+    // TODO implement other connector types
+  }
+}
+
+export function WebLNProvider ({ children }) {
+  const [provider, setProvider] = useState(null)
+  const [info, setInfo] = useState(null)
+  const [balance, setBalance] = useState(0)
+
+  const initProvider = async (provider) => {
+    const WebLNProviders = await import('@getalby/bitcoin-connect').then((mod) => mod.WebLNProviders)
+    if (provider instanceof WebLNProviders.LnbitsWebLNProvider) {
+      const connector = 'lnbits'
+      // TODO
+      //   we're reaching into private parts here.
+      //   is there a better way to show to user which connection was used?
+      const { hostname: connection } = new URL(provider._instanceUrl)
+      setInfo(o => ({ ...o, connector, connection }))
+    }
+    setProvider(provider)
+    const info = await provider.getInfo()
+    setInfo(o => ({ ...o, ...info }))
+    const { balance } = await provider.getBalance()
+    setBalance(balance)
+  }
+
+  const clearProvider = () => {
+    setProvider(null)
+    setInfo(null)
+  }
+
+  useEffect(() => {
+    async function effect () {
+      const provider = await fetchWebLNProvider()
+      if (!provider) return
+      initProvider(provider)
+    }
+    effect().catch(console.error)
+  }, [])
+
+  // event listeners
+  useEffect(() => {
+    let onConnectedUnsub
+    let onDisconnectedUnsub
+    async function effect () {
+      const onConnected = await import('@getalby/bitcoin-connect-react').then((mod) => mod.onConnected)
+      onConnectedUnsub = onConnected(async (provider) => {
+        // TODO
+        //   why is this not fired on page load when there is a provider available in local storage?
```

Comment:
I think Bitcoin Connect initializes before this functionality executes. So you probably also need to check if Bitcoin Connect is already connected. There is an api function `isConnected` that you can use.

Reply:
Mhh, makes sense. But what do I do if it's already connected? Is it too late then?

Is it not possible to first this event when the event listener was attached and it wasn't fired before? I don't know enough about JS internals to answer this question myself.

Reply:
@ekzyis I think you just need to execute the same functionality in both cases:
- if it is already connected, you can access the provider with `await requestProvider` - then do something with the provider.
- otherwise, subscribe to the connected event (which again returns the provider), as you are already doing

Does that work for you?

Reply:
> if it is already connected, you can access the provider with await requestProvider - then do something with the provider.

Using `requestProvider` prompts to connect a WebLN provider - even if there was no WebLN provider connected before. I only want to access the provider if there was one connected before. And not prompt the user to connect one. I believe this is what `onConnected` would solve.

https://github.com/stackernews/stacker.news/assets/27162016/ed2d8d13-9aec-4402-8afa-c6670616bbad

Reply:
@ekzyis `requestProvider` will not prompt if you're already connected. So as long as you check `isConnected` first, there should be no issue.

Reply:
Ah, I see [your comment here](https://github.com/getAlby/bitcoin-connect/issues/176#issuecomment-1871721940) and this makes sense! Will try it out.